ufl/core/external_operator.py

```diff
@@ -0,0 +1,278 @@
+        self._action_args, self._arguments = self._extract_coeffs_and_args(arguments)
+
+        # Make the coefficient associated to the external operator
+        ref_coefficient = Coefficient(function_space, count=count)
+        self._ufl_function_space = ref_coefficient.ufl_function_space()
+
+        # Checks
+        if derivatives is not None:
+            if not isinstance(derivatives, tuple):
+                raise TypeError("Expecting a tuple for derivatives and not %s" % derivatives)
+            if not len(derivatives) == len(self.ufl_operands):
+                error("Expecting a size of %s for %s" % (len(self.ufl_operands), derivatives))
+
+            self.derivatives = derivatives
+            # If we have arguments, the appropriate function space has already been set up upstream
+            if len(arguments) == 0:
+                # In this block, we construct the function space on which lives the ExternalOperator
+                # accordingly to the derivatives taken, for instance in 2d:
+                #      for V = VectorFunctionSpace(...)
+                #      if u = Function(V); e = ExternalOperator(u, function_space=V)
+                #      then, e.ufl_shape = (2,)
+                #            dedu.ufl_shape = (2,2)
+                #            ...
+                # Therefore, for 'dedu' we need to construct the new TensorElement to end up with the appropriate
+                # function space and we also need to store the VectorElement corresponding to V since it is on this
+                # function_space that we will interpolate the operands.
+                s = ref_coefficient.ufl_shape
+                for i, e in enumerate(self.derivatives):
+                    s += self.ufl_operands[i].ufl_shape * e
+                original_function_space = self._make_function_space(s, sub_element=ref_coefficient.ufl_element())
+                ufl_shape = original_function_space.ufl_element().reference_value_shape()
+            else:
+                ufl_shape = ref_coefficient.ufl_element().reference_value_shape()
+                original_function_space = ref_coefficient.ufl_function_space()
+        else:
+            ufl_shape = ref_coefficient.ufl_element().reference_value_shape()
+            original_function_space = ref_coefficient.ufl_function_space()
+            self.derivatives = (0,) * len(self.ufl_operands)
+
+        if coefficient is None:
+            coefficient = Coefficient(original_function_space, count=count)
+        elif not isinstance(coefficient, Coefficient):
+            raise TypeError('Expecting a Coefficient and not %s', type(coefficient))
+        self._coefficient = coefficient
+        self.ufl_shape = ufl_shape
+        self._original_function_space = self.coefficient.ufl_function_space()
+
+        self._extop_dependencies = [self, ]
+
+        if self.derivatives == (0,) * len(self.ufl_operands):
+            self._extop_master = self
```

Comment:
UFL objects are traditionally a DAG, this puts cycles into the expressions, which is bad. Why do you need to track dependencies like this?

Reply:
First, because when form compiling is not hit because of caching we don’t create the derivatives of `p`. So that is why we attach the information about what was made out of `p` to the kernel (cf. https://github.com/firedrakeproject/ufl/blob/af22ffd3583fef39621a879eb707f9a1b5e7e34a/ufl/algorithms/compute_form_data.py#L368)

Second, because we want to guarantee consistency across the coefficients associated with the derivatives of `p`. If we reconstruct a derivative of `p` that has already been made, we don’t need to make a new_object we just go through the dependency and returns the one with the correct derivatives multi-index (cf. https://github.com/firedrakeproject/ufl/blob/e9ca185c60e7d8344b5ba61000d8a88a9d13a742/ufl/core/external_operator.py#L198).

Reply:
OK, I think I understand. However, there are a few things that need to be tidied up.

1. The convention is that names starting with `_` are implementation detail (private). But here they are really providing public API. So can you make that change and document things.

2. Can you break the reference cycles (probably using `weakref.ref` or `weakref.proxy`)? I think the "parent" external operator needs to hold strong references to the child operators, but the children only need to hold weakrefs to the parent.

3. What does this in-place modification do to things like stability of the signature? Is it actually _necessary_, or is it a caching-based "performance" optimisation? For example, does all of your implementation require that differentiating the same extop twice with the same derivative multiindices give you back something that is object-identical, or just compares equal?